File: C++/balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> data;
        
        traverse(root, data);
        
        return buildBST(data, 0, data.size() - 1);
    }
    
    void traverse(TreeNode* root, vector<int> &data) {
        if (root) {
            traverse(root->left, data);
            data.push_back(root->val);
            traverse(root->right, data);
        }
    }
    
    TreeNode* buildBST(vector<int> &data, int i, int j) {
        TreeNode* result = nullptr;
        if (i <= j) {
            int mid = (i + j) / 2;
            result = new TreeNode(data[mid], 
                                 buildBST(data, i, mid-1),
                                 buildBST(data, mid+1, j));
        }
        return result;
    }
};

File: Go/balance-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
	list := buildList(root, nil)
	return buildTree(list)
}

func buildList(root *TreeNode, list []*TreeNode) []*TreeNode {
	if root == nil {
		return list
	}
	list = buildList(root.Left, list)
	list = append(list, root)
	list = buildList(root.Right, list)
	return list
}

func buildTree(list []*TreeNode) *TreeNode {
	if len(list) == 0 {
		return nil
	}
	mid := len(list) / 2
	root := list[mid]
	root.Left = buildTree(list[:mid])
	root.Right = buildTree(list[mid+1:])
	return root
}